Src/PerformanceCollector/Filtering/Filter.cs

```diff
@@ -0,0 +1,596 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.Filtering
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.Linq;
+    using System.Linq.Expressions;
+    using System.Reflection;
+
+    /// <summary>
+    /// Filter determines whether a telemetry document matches the criterion.
+    /// </summary>
+    /// <typeparam name="TTelemetry">Type of telemetry documents.</typeparam>
+    internal class Filter<TTelemetry>
+    {
+        private const string FieldNameCustomDimensionsPrefix = "CustomDimensions.";
+
+        private const string FieldNameCustomMetricsPrefix = "CustomMetrics.";
+
+        private const string FieldNameAsterisk = "*";
+
+        private const string CustomMetricsPropertyName = "Metrics";
+
+        private const string CustomDimensionsPropertyName = "Properties";
+
+        private static readonly MethodInfo DoubleToStringMethodInfo = GetMethodInfo<double, string>(x => x.ToString(CultureInfo.InvariantCulture));
```

Comment:
This is not reflection. This MethodInfo is used to construct a MethodCallExpression which is then compiled for performance and executed at collection-time for each ITelemetry.